modules/webbrowser/interaction/cefinteractionhandler.cpp

```diff
@@ -0,0 +1,131 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <modules/webbrowser/interaction/cefinteractionhandler.h>
+#include <modules/webbrowser/interaction/cefkeyboardmapping.h>
+#include <inviwo/core/interaction/events/resizeevent.h>
+#include <inviwo/core/interaction/events/wheelevent.h>
+
+namespace inviwo {
+
+CEFInteractionHandler::CEFInteractionHandler(CefRefPtr<CefBrowserHost> host) : host_(host){};
+
+void CEFInteractionHandler::invokeEvent(Event* event) {
+    switch (event->hash()) {
+        case ResizeEvent::chash():
+            host_->WasResized();
+            break;
+        case KeyboardEvent::chash(): {
+            host_->SendKeyEvent(mapKeyEvent(static_cast<KeyboardEvent*>(event)));
+            event->markAsUsed();
+            break;
+        }
+        case MouseEvent::chash(): {
+            auto mouseEvent = static_cast<MouseEvent*>(event);
+            updateMouseStates(mouseEvent);
+            auto cefMouseEvent = mapMouseEvent(mouseEvent);
+            if (MouseState::Move & mouseEvent->state()) {
+                bool mouseLeave = false;
+                host_->SendMouseMoveEvent(cefMouseEvent, mouseLeave);
+            } else {
+                CefBrowserHost::MouseButtonType type;
+                if (mouseEvent->button() & MouseButton::Left) {
+                    type = MBT_LEFT;
+                } else if (mouseEvent->button() & MouseButton::Middle) {
+                    type = MBT_MIDDLE;
+                } else {  // if (mouseEvent->button() & MouseButton::Right) {
+                    type = MBT_RIGHT;
+                }
+                bool mouseUp = MouseState::Release & mouseEvent->state() ? true : false;
+                int clickCount = MouseState::DoubleClick & mouseEvent->state() ? 2 : 1;
+                host_->SendMouseClickEvent(cefMouseEvent, type, mouseUp, clickCount);
+            }
+            event->markAsUsed();
+            break;
+        }
+        case WheelEvent::chash(): {
+            auto wheelEvent = static_cast<WheelEvent*>(event);
+            auto cefMouseEvent = mapMouseEvent(wheelEvent);
+            host_->SendMouseWheelEvent(cefMouseEvent, static_cast<int>(wheelEvent->delta().x),
+                                       static_cast<int>(wheelEvent->delta().y));
+            event->markAsUsed();
+            break;
+        }
```

Comment:
what about touch events?

Reply:
Not supported by Cef when using offscreen rendering to my knowledge. Perhaps we can inject them as mouse events anyway